### PR TITLE
Active response endpoint for new API

### LINKED ISSUE #2836: Implement active-response endpoint in new API
Hi team,

It is necessary to implement the endpoint for `active-response` in the new API.

Best regards,

Demetrio.

### PR DESCRIPTION
**Pull request by druizz90**
Hi team,

This PR closes #2836. It implements the endpoint for `active-response`. Below there is an example of an API call of `active-response` (`command` is executed properly):

Principal roadmap #2640.

```bash
# curl -X PUT -d '{"command":"restart-ossec0", "arguments": ["-", "null", "(from_the_server)", "(no_rule_id)"]}' -H 'Content-Type:application/json' "http://localhost:55000/active-response/000"
{
  "data": "Command sent.",
  "error": 0
}
```

Best regards,

Demetrio.

### CODE REVIEW

File: api/api/controllers/active_response_controller.py [hunk 1]
```diff
@@ -1,7 +1,58 @@
...
+    :type arguments: str
+
+    :rtype: dict
+    """
+
+    # get body parameters
+    if connexion.request.is_json:
+        active_response_model = ActiveResponse.from_dict(connexion.request.get_json())
+    else:
+        return 'ERROR', 400
```
Comment (crd1985) [inline api/api/controllers/active_response_controller.py]: This error message is not acceptable. We should tell the user why the input body is wrong and give a URL to doc.

File: api/api/controllers/active_response_controller.py [hunk 2]
```diff
@@ -1,7 +1,58 @@
...
+                          wait_for_complete=wait_for_complete,
+                          pretty=pretty,
+                          logger=logger
+                          )
+
+    data = loop.run_until_complete(dapi.distribute_function())
 
-def active_response():
-    pass
+    return data, 200
```
Comment (crd1985) [inline api/api/controllers/active_response_controller.py]: `data` should be a Model subclass.

File: api/api/models/active_response_model.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        """
+        return self._command
+
+    @arguments.setter
+    def arguments(self, arguments):
+        """Command running in the agent.
+
+        :param arguments: Command arguments
+        """
+        self._arguments = arguments
```
Comment (crd1985) [inline api/api/models/active_response_model.py]: No endline

**Comment by AdriiiPRodri** [comment]
# Status update
## Summary
This development is under continuous changes due to its dependence. 
I have improved the control of bugs and corrected several bugs in the endpoint, also improved the messages to make them more user friendly and useful.
Now you can call all agents with 000 or with all.

## Tasks
* [1.0h] Error Correction
* [2.5h] Exchange previous errors for new ones
* [2.5h] Exception control in the controller

## Stoppers
* Problems handling exceptions launched in the framework, thanks @crd1985 and @druizz90 

## Difficulties
* Exception capture

## Tests performed
### Fixed way to send command to all
```bash
root@07ad3748657a:/# curl -X PUT -d '{"command":"restart-ossec0", "arguments": ["-", "null", "(from_the_server)", "(no_rule_id)"]}' -H 'Content-Type:application/json' "http://localhost:55000/active-response/all"
{
  "data": null,
  "error": null,
  "message": "Command sent."
}
```
### Without body agent_id:
```bash
root@07ad3748657a:/# curl -X PUT "http://localhost:55000/active-response/all"
{
  "code": 1650,
  "dapi_errors": {},
  "detail": "Bad arguments",
  "remediation": "Please, visit [official documentation](https://documentation.wazuh.com/current/user-manual/api/reference.html#active-response) to get more information about how runs an Active Response command on a specified agent",
  "status": 400,
  "title": "Wazuh Error",
  "type": "about:blank"
}
```
### Malformed body:
```bash
root@07ad3748657a:/# curl -X PUT -d '{"commmand":"restart-ossec0", "arguments": ["-", "null", "(from_the_server)", "(no_rule_id)"]}' -H 'Content-Type:application/json' "http://localhost:55000/active-response/001"
{
  "detail": "'command' is a required property",
  "status": 400,
  "title": "Bad Request",
  "type": "about:blank"
}
```
### Agent not active:
```bash
root@07ad3748657a:/# curl -X PUT -d '{"command":"restart-ossec0", "arguments": ["-", "null", "(from_the_server)", "(no_rule_id)"]}' -H 'Content-Type:application/json' "http://localhost:55000/active-response/001"
{
  "code": 1651,
  "dapi_errors": {
    "master-node": {
      "error": "Agent is not active",
      "logfile": "WAZUH_HOME/logs/api.log"
    }
  },
  "detail": "Agent is not active",
  "remediation": "Please make sure the selected agent is active.",
  "status": 400,
  "title": "Wazuh Error",
  "type": "about:blank"
}
```
### With invalid agent_id:
```bash
root@07ad3748657a:/# curl -X PUT -d '{"command":"restart-ossec0", "arguments": ["-", "null", "(from_the_server)", "(no_rule_id)"]}' -H 'Content-Type:application/json' "http://localhost:55000/active-response/0001"
{
  "detail": "'0001' is too long\n\nFailed validating 'maxLength' in schema:\n    {'description': 'Agent ID.',\n     'maxLength': 3,\n     'minLength': 3,\n     'type': 'string',\n     'x-scope': ['', '#/components/parameters/agent_id']}\n\nOn instance:\n    '0001'",
  "status": 400,
  "title": "Bad Request",
  "type": "about:blank"
}
```
## Pending
* Do the unittests

**Comment by AdriiiPRodri** [comment]
# Status update
## Summary
I added the integrated tests to this endpoint, all correct.
## Tasks
* [1.0h] Integrated tests
## Difficulties
* Some returned values

## Tests performed
### Integrated test in api/api/test/test_active_response_endpoint.tavern.yaml
```bash
root@003bba3ba5df:/# /var/ossec/framework/python/bin/pytest /var/ossec/framework/python/lib/python3.7/site-packages/api/test/test_active_response_endpoint.tavern.yaml 
======================================================================== test session starts =========================================================================
platform linux -- Python 3.7.2, pytest-4.4.0, py-1.8.0, pluggy-0.9.0
rootdir: /var/ossec/framework/python/lib/python3.7/site-packages/api/test
plugins: tavern-0.26.3
collected 8 items                                                                                                                                                    

var/ossec/framework/python/lib/python3.7/site-packages/api/test/test_active_response_endpoint.tavern.yaml ........

========================================================================== warnings summary ==========================================================================
var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================================================ 8 passed, 1 warnings in 0.28 seconds ================================================================
```

**Comment by AdriiiPRodri** [comment]
## Test results
### Integrated tests
```bash
root@0aca173c7495:/var/ossec/framework/python/lib/python3.7/site-packages# /var/ossec/framework/python/bin/pytest -vv /var/ossec/framework/python/bin/test/test_active_response_endpoints.tavern.yaml 
============================== test session starts ==============================
platform linux -- Python 3.7.2, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /var/ossec/framework/python
plugins: tavern-0.26.4
collected 3 items                                                               

../../../bin/test/test_active_response_endpoints.tavern.yaml::GET authentication token PASSED [ 33%]
../../../bin/test/test_active_response_endpoints.tavern.yaml::PUT /active-response/{:agent_id} PASSED [ 66%]
../../../bin/test/test_active_response_endpoints.tavern.yaml::PUT /active-response/all PASSED [100%]

=============================== warnings summary ================================
yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

bin/test/test_active_response_endpoints.tavern.yaml::GET authentication token
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
===================== 3 passed, 2 warnings in 0.60 seconds ======================
```

### Unittests
```bash
root@0aca173c7495:/var/ossec/framework/python/lib/python3.7/site-packages/wazuh-3
.10.0-py3.7.egg/wazuh/tests# /var/ossec/framework/python/bin/pytest test_active_response.py 
============================== test session starts ==============================
platform linux -- Python 3.7.2, pytest-4.6.3, py-1.8.0, pluggy-0.12.0
rootdir: /var/ossec/framework/python/lib/python3.7/site-packages/wazuh-3.10.0-py3.7.egg/wazuh/tests
plugins: tavern-0.26.4
collected 6 items                                                               

test_active_response.py ......                                            [100%]

=========================== 6 passed in 0.13 seconds ============================
```

**Comment by davidjiglesias** [comment]
Updated tests

**Integration tests:**
```
============================================= test session starts =============================================
platform linux -- Python 3.7.2, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /var/ossec/framework/python/bin/python3.7
cachedir: .pytest_cache
rootdir: /var/ossec/framework/python/bin
plugins: cov-2.7.1, tavern-0.26.4
collected 3 items                                                                                             

test/test_active_response_endpoints.tavern.yaml::GET authentication token PASSED                        [ 33%]
test/test_active_response_endpoints.tavern.yaml::PUT /active-response/{:agent_id} PASSED                [ 66%]
test/test_active_response_endpoints.tavern.yaml::PUT /active-response PASSED                            [100%]

============================================== warnings summary ===============================================
/var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126
  /var/ossec/framework/python/lib/python3.7/site-packages/yaml/constructor.py:126: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    if not isinstance(key, collections.Hashable):

test/test_active_response_endpoints.tavern.yaml::GET authentication token
  /var/ossec/framework/python/lib/python3.7/site-packages/jwt/api_jwt.py:81: DeprecationWarning: It is strongly recommended that you pass in a value for the "algorithms" argument when calling decode(). This argument will be mandatory in a future version.
    DeprecationWarning

-- Docs: https://docs.pytest.org/en/latest/warnings.html
==================================== 3 passed, 2 warnings in 0.59 seconds =====================================
```
**Unit tests:**
```
============================= test session starts ==============================
platform linux -- Python 3.7.2, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: /opt/project/frameworkcollected 11 items

test_active_response.py ...........                                      [100%]

========================== 11 passed in 0.11 seconds ===========================
```

**Pull request closed without merging**